File: tests/go/tests-go-testing-allure3-playwright/tests/infra/config_reader_test.go
package infra_test

import (
	"strings"
	"testing"

	tests "tests-go-testing-allure3-playwright"

	allure "github.com/allure-framework/allure-go/commons/gotest"
	"github.com/stretchr/testify/require"
)

func wrapConfig(t *testing.T, name string, body func(*allure.Context)) {
	t.Helper()
	tests.Wrap(t, name, body, tests.LayerInfra("ConfigReader", "Test infra", "ConfigReader", "normal")...)
}

func TestResolveBaseURLAddsTrailingSlash(t *testing.T) {
	wrapConfig(t, "resolveBaseUrl adds trailing slash to HTTP baseUrl", func(a *allure.Context) {
		url, err := tests.ResolveBaseURL(tests.TestConfig{BaseURL: "http://localhost:3000"})
		require.NoError(t, err)
		require.Equal(t, "http://localhost:3000/", url)
	})
}

func TestResolveBaseURLKeepsTrailingSlash(t *testing.T) {
	wrapConfig(t, "resolveBaseUrl keeps trailing slash on baseUrl", func(a *allure.Context) {
		url, err := tests.ResolveBaseURL(tests.TestConfig{BaseURL: "http://localhost:3000/"})
		require.NoError(t, err)
		require.Equal(t, "http://localhost:3000/", url)
	})
}

func TestResolveBaseURLFailsWhenEmpty(t *testing.T) {
	wrapConfig(t, "resolveBaseUrl fails fast when baseUrl is empty", func(a *allure.Context) {
		_, err := tests.ResolveBaseURL(tests.TestConfig{BaseURL: ""})
		require.Error(t, err)
		require.Contains(t, err.Error(), "Set baseUrl")
	})
}

func TestResolveAPIBaseURLAddsTrailingSlash(t *testing.T) {
	wrapConfig(t, "resolveApiBaseUrl adds trailing slash to HTTP apiBaseUrl", func(a *allure.Context) {
		url, err := tests.ResolveAPIBaseURL(tests.TestConfig{APIBaseURL: "http://api.example.com"})
		require.NoError(t, err)
		require.Equal(t, "http://api.example.com/", url)
	})
}

func TestResolveAPIBaseURLFailsWhenEmpty(t *testing.T) {
	wrapConfig(t, "resolveApiBaseUrl fails fast when apiBaseUrl is empty", func(a *allure.Context) {
		_, err := tests.ResolveAPIBaseURL(tests.TestConfig{APIBaseURL: ""})
		require.Error(t, err)
		require.Contains(t, err.Error(), "Set apiBaseUrl")
	})
}

func TestLoadedBaseURLHasNoTrailingSlash(t *testing.T) {
	wrapConfig(t, "loaded baseUrl has no trailing slash (Owner file; Ui.open uses resolveBaseUrl)", func(a *allure.Context) {
		t.Setenv("STAND", "ci")
		t.Setenv("ENV", "ci")
		t.Setenv("BASE_URL", "")
		t.Setenv("API_BASE_URL", "")
		cfg := tests.LoadConfig()
		require.Equal(t, "http://localhost:9821", cfg.BaseURL)
		require.False(t, strings.HasSuffix(cfg.BaseURL, "/"))
	})
}

func TestResolveBaseURLUsesLoadedConfig(t *testing.T) {
	wrapConfig(t, "resolveBaseUrl uses loaded config", func(a *allure.Context) {
		t.Setenv("STAND", "ci")
		t.Setenv("ENV", "ci")
		t.Setenv("BASE_URL", "")
		t.Setenv("API_BASE_URL", "")
		url, err := tests.ResolveBaseURL(tests.LoadConfig())
		require.NoError(t, err)
		require.Equal(t, "http://localhost:9821/", url)
	})
}

func TestResolveAPIBaseURLUsesLoadedConfig(t *testing.T) {
	wrapConfig(t, "resolveApiBaseUrl uses loaded config", func(a *allure.Context) {
		t.Setenv("STAND", "ci")
		t.Setenv("ENV", "ci")
		t.Setenv("BASE_URL", "")
		t.Setenv("API_BASE_URL", "")
		url, err := tests.ResolveAPIBaseURL(tests.LoadConfig())
		require.NoError(t, err)
		require.Equal(t, "http://localhost:8800/", url)
	})
}

func TestPrivateConstructorKeepsUtilityClassClosed(t *testing.T) {
	wrapConfig(t, "private constructor keeps utility class closed", func(a *allure.Context) {
		require.NotNil(t, tests.ClosedConfigReader())
	})
}

File: tests/csharp/tests-csharp-xunit-api_request-playwright/Tests/Infra/ConfigReaderTest.cs
using System.Reflection;
using Allure.Net.Commons;
using Allure.Net.Commons.Attributes;
using Config;
using Tests;

namespace Tests.Infra;

[AllureLabel("layer", "infra")]
[AllureEpic("Test infra")]
[AllureFeature("ConfigReader")]
[AllureSeverity(SeverityLevel.normal)]
[Trait("TestCategory", "infra")]
[Trait("TestCategory", "infra_backend")]
[AllureSuite("ConfigReader")]
public sealed class ConfigReaderTest : AllureMeta
{
    [Fact(DisplayName = "resolveBaseUrl adds trailing slash to HTTP baseUrl")]
    public void ResolveBaseUrlAddsTrailingSlash()
    {
        var config = ConfigReader.ConfigWith(new Dictionary<string, string> { ["baseUrl"] = "http://localhost:3000" });
        Assert.Equal("http://localhost:3000/", ConfigReader.ResolveBaseUrl(config));
    }

    [Fact(DisplayName = "resolveBaseUrl keeps trailing slash on baseUrl")]
    public void ResolveBaseUrlKeepsTrailingSlash()
    {
        var config = ConfigReader.ConfigWith(new Dictionary<string, string> { ["baseUrl"] = "http://localhost:3000/" });
        Assert.Equal("http://localhost:3000/", ConfigReader.ResolveBaseUrl(config));
    }

    [Fact(DisplayName = "resolveBaseUrl fails fast when baseUrl is empty")]
    public void ResolveBaseUrlFailsWhenEmpty()
    {
        var config = ConfigReader.ConfigWith(new Dictionary<string, string> { ["baseUrl"] = "" });
        var error = Assert.Throws<InvalidOperationException>(() => ConfigReader.ResolveBaseUrl(config));
        Assert.Contains("Set baseUrl", error!.Message);
    }

    [Fact(DisplayName = "resolveApiBaseUrl adds trailing slash to HTTP apiBaseUrl")]
    public void ResolveApiBaseUrlAddsTrailingSlash()
    {
        var config = ConfigReader.ConfigWith(new Dictionary<string, string> { ["apiBaseUrl"] = "http://api.example.com" });
        Assert.Equal("http://api.example.com/", ConfigReader.ResolveApiBaseUrl(config));
    }

    [Fact(DisplayName = "resolveApiBaseUrl fails fast when apiBaseUrl is empty")]
    public void ResolveApiBaseUrlFailsWhenEmpty()
    {
        var config = ConfigReader.ConfigWith(new Dictionary<string, string> { ["apiBaseUrl"] = "" });
        var error = Assert.Throws<InvalidOperationException>(() => ConfigReader.ResolveApiBaseUrl(config));
        Assert.Contains("Set apiBaseUrl", error!.Message);
    }

    [Fact(DisplayName = "loaded baseUrl has no trailing slash (Owner file; Ui.open uses resolveBaseUrl)")]
    public void LoadedBaseUrlHasNoTrailingSlash()
    {
        UsingCiStand(() =>
        {
            Assert.Equal("http://localhost:9821", ConfigReader.TestConfig.BaseUrl);
        });
    }

    [Fact(DisplayName = "resolveBaseUrl uses loaded config")]
    public void ResolveBaseUrlUsesLoadedConfig()
    {
        UsingCiStand(() =>
        {
            Assert.Equal("http://localhost:9821/", ConfigReader.ResolveBaseUrl());
        });
    }

    [Fact(DisplayName = "resolveApiBaseUrl uses loaded config")]
    public void ResolveApiBaseUrlUsesLoadedConfig()
    {
        UsingCiStand(() =>
        {
            Assert.Equal("http://localhost:8800/", ConfigReader.ResolveApiBaseUrl());
        });
    }

    [Fact(DisplayName = "private constructor keeps utility class closed")]
    public void PrivateConstructorIsReachable()
    {
        var ctor = typeof(ConfigReader).GetConstructor(BindingFlags.Instance | BindingFlags.NonPublic, Type.EmptyTypes);
        Assert.NotNull(ctor);
        Assert.NotNull(ctor!.Invoke(null));
        Assert.NotNull(ConfigReader.ClosedConfigReader());
    }

    private static void UsingCiStand(Action body)
    {
        var previous = Snapshot(["STAND", "ENV", "BASE_URL", "API_BASE_URL"]);
        try
        {
            Environment.SetEnvironmentVariable("STAND", "ci");
            Environment.SetEnvironmentVariable("ENV", "ci");
            Environment.SetEnvironmentVariable("BASE_URL", "");
            Environment.SetEnvironmentVariable("API_BASE_URL", "");
            body();
        }
        finally
        {
            Restore(previous);
        }
    }

    private static Dictionary<string, string?> Snapshot(IEnumerable<string> keys) =>
        keys.ToDictionary(key => key, Environment.GetEnvironmentVariable);

    private static void Restore(Dictionary<string, string?> previous)
    {
        foreach (var (key, value) in previous)
        {
            Environment.SetEnvironmentVariable(key, value);
        }
    }
}
